# Housing Rental Analysis for San Francisco

In this challenge, your job is to use your data visualization skills, including aggregation, interactive visualizations, and geospatial analysis, to find properties in the San Francisco market that are viable investment opportunities.

## Instructions

Use the `san_francisco_housing.ipynb` notebook to visualize and analyze the real-estate data.

Note that this assignment requires you to create a visualization by using hvPlot and GeoViews. Additionally, you need to read the `sfo_neighborhoods_census_data.csv` file from the `Resources` folder into the notebook and create the DataFrame that you’ll use in the analysis.

The main task in this Challenge is to visualize and analyze the real-estate data in your Jupyter notebook. Use the `san_francisco_housing.ipynb` notebook to complete the following tasks:

* Calculate and plot the housing units per year.

* Calculate and plot the average prices per square foot.

* Compare the average prices by neighborhood.

* Build an interactive neighborhood map.

* Compose your data story.

### Calculate and Plot the Housing Units per Year

For this part of the assignment, use numerical and visual aggregation to calculate the number of housing units per year, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

2. Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

3. Style and format the line plot to ensure a professionally styled visualization.

4. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting bar chart.](Images/zoomed-housing-units-by-year.png)

5. Answer the following question:

    * What’s the overall trend in housing units over the period that you’re analyzing?

### Calculate and Plot the Average Sale Prices per Square Foot

For this part of the assignment, use numerical and visual aggregation to calculate the average prices per square foot, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Group the data by year, and then average the results. What’s the lowest gross rent that’s reported for the years that the DataFrame includes?

2. Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

3. Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

    > **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/avg-sale-px-sq-foot-gross-rent.png)

6. Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

    * Did any year experience a drop in the average sale price per square foot compared to the previous year?

    * If so, did the gross rent increase or decrease during that year?

### Compare the Average Sale Prices by Neighborhood

For this part of the assignment, use interactive visualizations and widgets to explore the average sale price per square foot by neighborhood. To do so, complete the following steps:

1. Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

2. Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

3. Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/pricing-info-by-neighborhood.png)

6. Use the interactive visualization to answer the following question:

    * For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 

### Build an Interactive Neighborhood Map

For this part of the assignment, explore the geospatial relationships in the data by using interactive visualizations with hvPlot and GeoViews. To build your map, use the `sfo_data_df` DataFrame (created during the initial import), which includes the neighborhood location data with the average prices. To do all this, complete the following steps:

1. Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

2. Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

3. Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. The second cell cleans the data and sets the “Neighborhood” column. Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

4. Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

    * Set the `geo` parameter to True.
    * Set the `size` parameter to “sale_price_sqr_foot”.
    * Set the `color` parameter to “gross_rent”.
    * Set the `frame_width` parameter to 700.
    * Set the `frame_height` parameter to 500.
    * Include a descriptive title.

Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of a scatter plot created with hvPlot and GeoViews.](Images/6-4-geoviews-plot.png)

5. Use the interactive map to answer the following question:

    * Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?

### Compose Your Data Story

Based on the visualizations that you created, answer the following questions:

* How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighborhoods across San Francisco?

* What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighborhoods exist that you would suggest for investment, and why?

In [78]:
# Import the required libraries and dependencies
import pandas as pd
import hvplot.pandas
from pathlib import Path

## Import the data 

In [79]:
# Using the read_csv function and Path module, create a DataFrame 
# by importing the sfo_neighborhoods_census_data.csv file from the Resources folder
sfo_data_df = Path(r"C:\Users\reide\Downloads\sfo_neighborhoods_census_data.csv")
sfo = pd.read_csv(sfo_data_df, index_col="year", infer_datetime_format=True, parse_dates=True)

# Review the first five rows of df #

sfo.iloc[0:5].head()

neighborhood  sale_price_sqr_foot  housing_units  gross_rent
year                                                                        
2010-01-01      Alamo Square           291.182945         372560        1239
2010-01-01        Anza Vista           267.932583         372560        1239
2010-01-01           Bayview           170.098665         372560        1239
2010-01-01  Buena Vista Park           347.394919         372560        1239
2010-01-01  Central Richmond           319.027623         372560        1239

In [80]:
#last five rows of the DataFrame#

sfo.iloc[392:397].head()

neighborhood  sale_price_sqr_foot  housing_units  \
year                                                                     
2016-01-01          Telegraph Hill           903.049771         384242   
2016-01-01              Twin Peaks           970.085470         384242   
2016-01-01  Van Ness/ Civic Center           552.602567         384242   
2016-01-01       Visitacion Valley           328.319007         384242   
2016-01-01           Westwood Park           631.195426         384242   

            gross_rent  
year                    
2016-01-01        4390  
2016-01-01        4390  
2016-01-01        4390  
2016-01-01        4390  
2016-01-01        4390

---

## Calculate and Plot the Housing Units per Year

For this part of the assignment, use numerical and visual aggregation to calculate the number of housing units per year, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

2. Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

3. Style and format the line plot to ensure a professionally styled visualization.

4. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting bar chart.](Images/zoomed-housing-units-by-year.png)

5. Answer the following question:

    * What’s the overall trend in housing units over the period that you’re analyzing?



### Step 1: Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

In [81]:
# Create a numerical aggregation that groups the data by the year and then averages the results#
housing_units_by_year = sfo.groupby(["year","housing_units"]).mean()

# Review the DataFrame # 

housing_units_by_year

sale_price_sqr_foot  gross_rent
year       housing_units                                 
2010-01-01 372560                  369.344353      1239.0
2011-01-01 374507                  341.903429      1530.0
2012-01-01 376454                  399.389968      2324.0
2013-01-01 378401                  483.600304      2971.0
2014-01-01 380348                  556.277273      3528.0
2015-01-01 382295                  632.540352      3739.0
2016-01-01 384242                  697.643709      4390.0

### Step 2: Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

### Step 3: Style and format the line plot to ensure a professionally styled visualization.

In [82]:
# Create a visual aggregation explore the housing units by year#

housing_units_by_year.hvplot.bar(
    title='Housing Units in San Francisco from 2010 to 2016',
    x="year",
    y="housing_units",
    yformatter = '%.3f',
    
)

:Bars   [year]   (housing_units)

### Step 5: Answer the following question:

**Question:** What is the overall trend in housing_units over the period being analyzed?

**Answer:** #

The overall trend we can see in the housing market is that the number of housing units is increasing but not by a very a large margin. So we can see that over the course of 6 years the number of housing units has increased by about 12000 units. 12000 unit increase for a city with the growth that San Fransisco has had during that time frame is likely not enough to be able to house all of the new people who would have come into the city during that time.

---

## Calculate and Plot the Average Sale Prices per Square Foot

For this part of the assignment, use numerical and visual aggregation to calculate the average prices per square foot, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Group the data by year, and then average the results. What’s the lowest gross rent that’s reported for the years that the DataFrame includes?

2. Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

3. Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

    > **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/avg-sale-px-sq-foot-gross-rent.png)

6. Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

    * Did any year experience a drop in the average sale price per square foot compared to the previous year?

    * If so, did the gross rent increase or decrease during that year?



### Step 1: Group the data by year, and then average the results.

In [83]:
# Create a numerical aggregation by grouping the data by year and averaging the results
prices_square_foot_by_year = sfo.groupby(["year"]).mean()

# Review the resulting DataFrame# 

prices_square_foot_by_year


sale_price_sqr_foot  housing_units  gross_rent
year                                                      
2010-01-01           369.344353       372560.0      1239.0
2011-01-01           341.903429       374507.0      1530.0
2012-01-01           399.389968       376454.0      2324.0
2013-01-01           483.600304       378401.0      2971.0
2014-01-01           556.277273       380348.0      3528.0
2015-01-01           632.540352       382295.0      3739.0
2016-01-01           697.643709       384242.0      4390.0

In [84]:
prices_square_foot_by_year.min()

sale_price_sqr_foot       341.903429
housing_units          372560.000000
gross_rent               1239.000000
dtype: float64

**Question:** What is the lowest gross rent reported for the years included in the DataFrame?

**Answer:** #

The lowest gross rent reported for the years included in the datafame is $1239

### Step 2: Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

In [85]:
# Filter out the housing_units column, creating a new DataFrame 
# Keep only sale_price_sqr_foot and gross_rent averages per year

prices_square_foot_by_year = sfo[["sale_price_sqr_foot", "gross_rent"]]

# Review the DataFrame#

prices_square_foot_by_year.groupby(['year']).mean()

sale_price_sqr_foot  gross_rent
year                                       
2010-01-01           369.344353      1239.0
2011-01-01           341.903429      1530.0
2012-01-01           399.389968      2324.0
2013-01-01           483.600304      2971.0
2014-01-01           556.277273      3528.0
2015-01-01           632.540352      3739.0
2016-01-01           697.643709      4390.0

### Step 3: Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`

### Step 4: Style and format the line plot to ensure a professionally styled visualization.


In [86]:
# Plot prices_square_foot_by_year. 
# Inclued labels for the x- and y-axes, and a title#


prices_square_foot_by_year.hvplot.line(
     x="year",
    y=["gross_rent", "sale_price_sqr_foot"],
    ylabel= "Gross Rent / Sale Price Per Square Foot",
    title="Sale Price Per Square Foot and Average Gross Rent - 2010-2016 - San Francisco",
)

:NdOverlay   [Variable]
   :Curve   [year]   (value)

### Step 6: Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

**Question:** Did any year experience a drop in the average sale price per square foot compared to the previous year?

**Answer:** # Between 2010-2011 we can see a drop in the average sale price per sq ft from 369 in 2010 to 341 in 2011

**Question:** If so, did the gross rent increase or decrease during that year?

**Answer:** # The gross rent in that year increased

---

## Compare the Average Sale Prices by Neighborhood

For this part of the assignment, use interactive visualizations and widgets to explore the average sale price per square foot by neighborhood. To do so, complete the following steps:

1. Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

2. Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

3. Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/pricing-info-by-neighborhood.png)

6. Use the interactive visualization to answer the following question:

    * For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 


### Step 1: Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

In [87]:
# Group by year and neighborhood and then create a new dataframe of the mean values
prices_by_year_by_neighborhood = sfo.groupby(['year', 'neighborhood']).mean()
# Review the DataFrame

prices_by_year_by_neighborhood

# YOUR CODE HERE

sale_price_sqr_foot  housing_units  \
year       neighborhood                                                 
2010-01-01 Alamo Square                     291.182945       372560.0   
           Anza Vista                       267.932583       372560.0   
           Bayview                          170.098665       372560.0   
           Buena Vista Park                 347.394919       372560.0   
           Central Richmond                 319.027623       372560.0   
...                                                ...            ...   
2016-01-01 Telegraph Hill                   903.049771       384242.0   
           Twin Peaks                       970.085470       384242.0   
           Van Ness/ Civic Center           552.602567       384242.0   
           Visitacion Valley                328.319007       384242.0   
           Westwood Park                    631.195426       384242.0   

                                   gross_rent  
year       neighborhood                        
2010-01-01 Alamo Square                1239.0  
           Anza Vista                  1239.0  
           Bayview                     1239.0  
           Buena Vista Park            1239.0  
           Central Richmond            1239.0  
...                                       ...  
2016-01-01 Telegraph Hill              4390.0  
           Twin Peaks                  4390.0  
           Van Ness/ Civic Center      4390.0  
           Visitacion Valley           4390.0  
           Westwood Park               4390.0  

[397 rows x 3 columns]

### Step 2: Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

In [88]:
# Filter out the housing_units
prices_by_year_by_neighborhood = sfo[["sale_price_sqr_foot", "gross_rent", "neighborhood"]]

# Review the first and last five rows of the DataFrame

prices_by_year_by_neighborhood.iloc[0:5]

sale_price_sqr_foot  gross_rent      neighborhood
year                                                         
2010-01-01           291.182945        1239      Alamo Square
2010-01-01           267.932583        1239        Anza Vista
2010-01-01           170.098665        1239           Bayview
2010-01-01           347.394919        1239  Buena Vista Park
2010-01-01           319.027623        1239  Central Richmond

In [89]:
#last five rows of the DataFrame#

prices_by_year_by_neighborhood.iloc[392:397]

sale_price_sqr_foot  gross_rent            neighborhood
year                                                               
2016-01-01           903.049771        4390          Telegraph Hill
2016-01-01           970.085470        4390              Twin Peaks
2016-01-01           552.602567        4390  Van Ness/ Civic Center
2016-01-01           328.319007        4390       Visitacion Valley
2016-01-01           631.195426        4390           Westwood Park

### Step 3: Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

### Step 4: Style and format the line plot to ensure a professionally styled visualization.

In [90]:
# Use hvplot to create an interactive line plot of the average price per square foot
# The plot should have a dropdown selector for the neighborhood
# YOUR CODE HERE

prices_by_year_by_neighborhood.hvplot.line(
    groupby="neighborhood",
    x="year",
    y=["gross_rent", "sale_price_sqr_foot"],
    ylabel= "gross rent and sale price per square foot",
    title="Sale Price Per Square Foot and Average Gross Rent - 2010-2016 - San Francisco"
)

:DynamicMap   [neighborhood]
   :NdOverlay   [Variable]
      :Curve   [year]   (value)

### Step 6: Use the interactive visualization to answer the following question:

**Question:** For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 

**Answer:** For Anza Vista neighborhood, the average sale price per square foot in 2016 is less than that what is listed for 2012

---

## Build an Interactive Neighborhood Map

For this part of the assignment, explore the geospatial relationships in the data by using interactive visualizations with hvPlot and GeoViews. To build your map, use the `sfo_data_df` DataFrame (created during the initial import), which includes the neighborhood location data with the average prices. To do all this, complete the following steps:

1. Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

2. Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

3. Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. The second cell cleans the data and sets the “Neighborhood” column. Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

4. Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

    * Set the `size` parameter to “sale_price_sqr_foot”.

    * Set the `color` parameter to “gross_rent”.

    * Set the `size_max` parameter to “25”.

    * Set the `zoom` parameter to “11”.

Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of a scatter plot created with hvPlot and GeoViews.](Images/6-4-geoviews-plot.png)

5. Use the interactive map to answer the following question:

    * Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?


### Step 1: Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

In [91]:
# Load neighborhoods coordinates data
neighborhood_locations_df =  Path(r"C:\Users\reide\Downloads\pyvizstuff\UTOR-VIRT-FIN-PT-11-2022-U-LOLC-main-06-PyViz\06-PyViz\Unit 6 Homework\Instructions\Starter_Code\Resources\neighborhoods_coordinates.csv")
Location = pd.read_csv(neighborhood_locations_df, index_col="Neighborhood", infer_datetime_format=True, parse_dates=True)

# Review the DataFrame
Location

Lat         Lon
Neighborhood                             
Alamo Square        37.791012 -122.402100
Anza Vista          37.779598 -122.443451
Bayview             37.734670 -122.401060
Bayview Heights     37.728740 -122.410980
Bernal Heights      37.728630 -122.443050
...                       ...         ...
West Portal         37.740260 -122.463880
Western Addition    37.792980 -122.435790
Westwood Highlands  37.734700 -122.456854
Westwood Park       37.734150 -122.457000
Yerba Buena         37.792980 -122.396360

[73 rows x 2 columns]

### Step 2: Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

In [92]:
# Calculate the mean values for each neighborhood
all_neighborhood_info_df = sfo.groupby(['neighborhood']).mean()

# Review the resulting DataFrame
all_neighborhood_info_df

sale_price_sqr_foot  housing_units   gross_rent
neighborhood                                                       
Alamo Square                 366.020712      378401.00  2817.285714
Anza Vista                   373.382198      379050.00  3031.833333
Bayview                      204.588623      376454.00  2318.400000
Bayview Heights              590.792839      382295.00  3739.000000
Bernal Heights               576.746488      379374.50  3080.333333
...                                 ...            ...          ...
West Portal                  498.488485      376940.75  2515.500000
Western Addition             307.562201      377427.50  2555.166667
Westwood Highlands           533.703935      376454.00  2250.500000
Westwood Park                687.087575      382295.00  3959.000000
Yerba Buena                  576.709848      377427.50  2555.166667

[73 rows x 3 columns]

### Step 3: Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. 

Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. 

The second cell cleans the data and sets the “Neighborhood” column. 

Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

In [93]:
# Using the Pandas `concat` function, join the 
# neighborhood_locations_df and the all_neighborhood_info_df DataFrame
# The axis of the concatenation is "columns".
# The concat function will automatially combine columns with
# identical information, while keeping the additional columns.
all_neighborhoods_df = pd.concat(
    [Location, all_neighborhood_info_df], 
    axis="columns",
    sort=False
)

# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())


Lat         Lon  sale_price_sqr_foot  housing_units  \
Alamo Square     37.791012 -122.402100           366.020712       378401.0   
Anza Vista       37.779598 -122.443451           373.382198       379050.0   
Bayview          37.734670 -122.401060           204.588623       376454.0   
Bayview Heights  37.728740 -122.410980           590.792839       382295.0   
Bernal Heights   37.728630 -122.443050                  NaN            NaN   

                  gross_rent  
Alamo Square     2817.285714  
Anza Vista       3031.833333  
Bayview          2318.400000  
Bayview Heights  3739.000000  
Bernal Heights           NaN

Lat        Lon  sale_price_sqr_foot  housing_units  \
Yerba Buena      37.79298 -122.39636           576.709848       377427.5   
Bernal Heights        NaN        NaN           576.746488       379374.5   
Downtown              NaN        NaN           391.434378       378401.0   
Ingleside             NaN        NaN           367.895144       377427.5   
Outer Richmond        NaN        NaN           473.900773       378401.0   

                  gross_rent  
Yerba Buena      2555.166667  
Bernal Heights   3080.333333  
Downtown         2817.285714  
Ingleside        2509.000000  
Outer Richmond   2817.285714

In [94]:
# Call the dropna function to remove any neighborhoods that do not have data
all_neighborhoods_df = all_neighborhoods_df.reset_index().dropna()

# Rename the "index" column as "Neighborhood" for use in the Visualization
all_neighborhoods_df = all_neighborhoods_df.rename(columns={"index": "Neighborhood"})

# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())

Neighborhood        Lat         Lon  sale_price_sqr_foot  \
0      Alamo Square  37.791012 -122.402100           366.020712   
1        Anza Vista  37.779598 -122.443451           373.382198   
2           Bayview  37.734670 -122.401060           204.588623   
3   Bayview Heights  37.728740 -122.410980           590.792839   
5  Buena Vista Park  37.768160 -122.439330           452.680591   

   housing_units   gross_rent  
0       378401.0  2817.285714  
1       379050.0  3031.833333  
2       376454.0  2318.400000  
3       382295.0  3739.000000  
5       378076.5  2698.833333

Neighborhood       Lat         Lon  sale_price_sqr_foot  \
68         West Portal  37.74026 -122.463880           498.488485   
69    Western Addition  37.79298 -122.435790           307.562201   
70  Westwood Highlands  37.73470 -122.456854           533.703935   
71       Westwood Park  37.73415 -122.457000           687.087575   
72         Yerba Buena  37.79298 -122.396360           576.709848   

    housing_units   gross_rent  
68      376940.75  2515.500000  
69      377427.50  2555.166667  
70      376454.00  2250.500000  
71      382295.00  3959.000000  
72      377427.50  2555.166667

### Step 4: Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

* Set the `geo` parameter to True.
* Set the `size` parameter to “sale_price_sqr_foot”.
* Set the `color` parameter to “gross_rent”.
* Set the `frame_width` parameter to 700.
* Set the `frame_height` parameter to 500.
* Include a descriptive title.

In [95]:
# Create a plot to analyze neighborhood info
###i am having issues with my conda path env some packages arent installing correctly
##like the ones required to generate the plot but the code below should generate the appropriate map


all_neighborhoods_df.hvplot.points('Lon','Lat', geo=True, size='sale_price_sqr_foot',
                                   color = 'gross_rent', frame_width=700, frame_height=500,
                                   title = 'San Francisco housing price/sq ft per neighborhood')

ImportError: Geographic projection support requires GeoViews, pyproj and cartopy.

### Step 5: Use the interactive map to answer the following question:

**Question:** Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?

**Answer:** # As per my code listed below and executed in the cell below as my map is not working the neighborhood with the highest gross rent is yerba buena and the
high sale price per square foot is also yerba buena

## Compose Your Data Story

Based on the visualizations that you have created, compose a data story that synthesizes your analysis by answering the following questions:

**Question:**  How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighborhoods across San Francisco?

**Answer:** The trend in rental income does not follow suite with the trend in price increase as the rental prices have sky rocketed in the past 10 years making the city
very unaffordale to live. This same trend holds true for most of the neighborhoods as it has a domino effect starting from within the most expensive areas becoming more and more expensive
the surrounding neighborhoods slowly become more and more expensive and so on pushing those who can't afford it futher and further away from the most expensive areas. 

**Question:** What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighborhoods exist that you would suggest for investment, and why?

**Answer:** # One insight i would share with the company is that because the prices are going to be so high in san francisco that although pricing is going to be heavily dependant on decision making when
it comes to renting a placce i do not think it will be the only factor - the app will have to have a slightly more "human" element in the form of neighborhood descriptions or non monetary score
such as nightlife, coffee, food, proximity of schools for families etc...

In [96]:
prices_by_year_by_neighborhood_maxgr = sfo[["gross_rent", "neighborhood"]].max()

In [97]:
prices_by_year_by_neighborhood_max

gross_rent             4390
neighborhood    Yerba Buena
dtype: object

In [98]:
prices_by_year_by_neighborhood_maxsqr = sfo[["sale_price_sqr_foot", "neighborhood"]].max()

In [99]:
prices_by_year_by_neighborhood_maxsqr

sale_price_sqr_foot    2258.702832
neighborhood           Yerba Buena
dtype: object